libcontainer/container_linux.go

```diff
@@ -836,6 +836,84 @@ func (c *linuxContainer) handleCriuConfigurationFile(rpcOpts *criurpc.CriuOpts)
 	}
 }
 
+func (c *linuxContainer) handleCheckpointingExternalNamespaces(rpcOpts *criurpc.CriuOpts, t configs.NamespaceType) error {
+	var nsType string
+	var minVersion int
+	switch configs.NEWNET {
+	case configs.NEWNET:
+		nsType = "Net"
+		// CRIU supports different external namespace with different released CRIU versions.
+		// For network namespaces to work we need at least criu 3.11.0 => 31100.
+		minVersion = 31100
+	default:
+		return nil
+	}
```

Comment:
Now it's very clear we're doing the same thing twice, so maybe have something like

```go
// Checks if external namespace t is supported by the current criu.
// If yes, returns the nsType key, otherwise, an empty string is returned.
func criuExtNsSupported(t configs.NamespaceType) string {
  var (
    key string
    // CRIU supports different external namespace with different released CRIU versions.
    ver int
  )
  switch t {
    case configs.NEWNET:
      key = "Net"
      ver = 31100
    default:
      return ""
  }
  if err := c.checkCriuVersion(minVersion); err != nil {
    return ""
  }

  return key
}
```

and use it from both `handleCheckpointingExternalNamespaces` and `handleRestoringExternalNamespaces`?

Can be a `map[configs.NamespaceType]struct{minVersion int, key string}` instead of switch, or whatever works.

In fact we already have configs.NsName(), the only issue is capitalization, so maybe use it and `strings.Title()`.

IE

```go
func criuSupportsExtNS(t configs.NamespaceType) bool {
  switch t {
    case configs.NEWNET:
      return c.checkCriuVersion(31100) == nil
    ...
  }
  return false
}

func criuNsToKey(t configs.NamespaceType) {
  return "ExtRoot" + strings.Title(configs.NsName(t) + "NS"
}

...

func (c *linuxContainer) handleCheckpointingExternalNamespaces(rpcOpts *criurpc.CriuOpts, t configs.NamespaceType) error {
  if !criuSupportsExtNs(t) {
    return nil
  }
  ...
  criuExternal := fmt.Sprintf("%s[%d]:%s", configs.NsName(t), ns.Ino, criuNsKey(t))
  ...
}

...
  inheritFd.Key = proto.String(criuNsKey(t))
```

Reply:
reworked like this